### PR TITLE
Add lightweight v2 provider interface starter

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit is a lighterweight v2 provider interface. This is an
alternative to what is built in #4885. While the model in #4885 is a
desireable end state it requires a lot of changes all at once for
providers and potentially users. Instead this commit brings the core
concept from #4885 of a cleaner explicitly versioned abstract interface
but minimizes the changes to the data model used in v1. Only some small
changes are made, mainly that jobs can be sync or async, Backend.run()
takes a circuit or schedule, options are set via an Options object at
__init__, with set_option(), or as kwargs on run(). In all other places
the object models from the v1 provider interface are used. This makes
the migration to a versioned interface simpler to start. From there we
can continue to evolve the interface as was done in #4485, like moving
to a target object, reimplementing properties and defaults as versioned
objects, etc.

Since the differences here are so small this commit brings the basicaer
provider over to the v2 provider interface. This can be done with
minimal effort to showcase how similar the 2 interfaces are.

### Details and comments

### CODE REVIEW

**Comment by mtreinish** [comment]
So this is blocked on aqua doing explicit type checking in its quantum instance for a `BaseBackend` class and Basic Aer's backends being a v2 `Backend` object. I think this means we'll have to split out the basic aer v2 conversion to post release. Then we can merge v2 provider support to aqua after this merges. Then there shouldn't be a blocker on this anymore.

**Comment by mtreinish** [comment]
I've removed the basicaer changes that switched the basic aer provider to be the first user of the v2 providers interface. This worksaround the tutorials job failure. I also pushed up https://github.com/Qiskit/qiskit-aqua/pull/1270 to add support for type checking v2 Backend classes to aqua. After this PR merges we can merge the aqua PR. Then after both terra and aqua release I'll push up a PR migrating basic aer to use the v2 interface (we can also start the migration of aer and ibmq too).

File: qiskit/providers/v2/options.py
Comment (jyu00) [inline qiskit/providers/v2/options.py:16]: If this inherits `SimpleNamespace` or even `dict`, then user can more easily list all available options and use autocomplete.

File: qiskit/providers/v2/options.py
Comment (jyu00) [inline qiskit/providers/v2/options.py:20]:
```suggestion
    on. The properties of the class are intended to be all dynamically
```

File: qiskit/providers/v2/options.py
Comment (jyu00) [inline qiskit/providers/v2/options.py:23]:
```suggestion
    that should be a configuration of the backend class itself instead of the
```

File: qiskit/providers/v2/job.py
Comment (jyu00) [inline qiskit/providers/v2/job.py:45]: This also needs `version = 1`

File: qiskit/providers/v2/provider.py
Comment (jyu00) [inline qiskit/providers/v2/provider.py:33]: This also needs `version = 1`

File: qiskit/providers/v2/options.py
Comment (ajavadia) [inline qiskit/providers/v2/options.py:30]: is it possible to make them into attributes of `Options` so updating them looks like `backend.options.shots=2000`?

File: qiskit/providers/v2/job.py
Comment (mtreinish) [reply qiskit/providers/v2/job.py:125]: It is used by basic aer: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/providers/basicaer/qasm_simulator.py#L393 I did rip it out in the v2 migration (https://github.com/Qiskit/qiskit-terra/pull/5086/commits/9ede763d57632d593549facdac3e02d8cb75dc3b is the revert of that) because I made basic aer hardcoded as synchronous.

Comment (mtreinish) [reply qiskit/providers/v2/job.py:125]: It looks like the code was ported to aer too: https://github.com/Qiskit/qiskit-aer/blob/8114479b0938f26f2f69f7e091368c3ec8e8cba1/qiskit/providers/aer/backends/aerbackend.py#L111 https://github.com/Qiskit/qiskit-aer/blob/9492a71633223c451a0bc606e41d898b2c2779c4/qiskit/providers/aer/backends/pulse_simulator.py#L134

File: qiskit/providers/v2/job.py
Comment (ajavadia) [inline qiskit/providers/v2/job.py:80]:
```suggestion
    def in_final_state(self) -> bool:
        """Return whether the job is in a final job state such as ``DONE`` or ``ERROR``."""
```

File: qiskit/providers/v2/backend.py
Comment (ajavadia) [inline qiskit/providers/v2/backend.py:206]: this sentence seems incomplete

Comment (ajavadia) [reply qiskit/providers/v2/backend.py:206]: do you mean a QasmQobj and PulseQobj will be supported for now? but the type in docstring doesn't say that

**Review by ajavadia — COMMENTED** [review]
i'm a little confused by v1 vs. v2 - there's a v2 folder, is the intention that there will be v3, v4, etc. folders in the future? or by v2 you mean "everything that will be versioned from now on"?

also ProviderV1 is under the v2 folder. If the idea is that ProviderV2, ProviderV3, etc. will all be in this folder, then I think this folder is a bit redundant and everything should be under the main provider folder.

File: qiskit/providers/v2/backend.py
Comment (mtreinish) [reply qiskit/providers/v2/backend.py:206]: Oops yeah I missed that. I'll fix it now. 

I didn't want to put `Qobj` in the docstring type because I didn't want to encourage its use for new providers. But, I still wanted to call out anyone migrating from v1 to v2 should probably still support Qobj (but deprecate it) objects for backwards compatibility reasons.

File: qiskit/providers/v2/options.py
Comment (mtreinish) [reply qiskit/providers/v2/options.py:30]: Yeah, @jyu00's suggestion above about using `SimpleNamespace` should enable that pattern for setting it. I'd still like to keep the method because it has the membership check with a custom exception.

**Comment by mtreinish** [comment]
> i'm a little confused by v1 vs. v2 - there's a v2 folder, is the intention that there will be v3, v4, etc. folders in the future? or by v2 you mean "everything that will be versioned from now on"?

The later, basically the v2 subdirectory was supposed to indicate the new version of the entire interface. In that new interface all the objects are individually versioned (starting at 1). I've been thinking of it as more of a `Backend` v2.1, etc. I just used ints for the object versions (1 vs 2.1) because it means we don't have to do string parsing on versioning negotiation, which isn't part of this PR since it's just the initial framework but will be a follow on when we start having more than one version.

> 
> also ProviderV1 is under the v2 folder. If the idea is that ProviderV2, ProviderV3, etc. will all be in this folder, then I think this folder is a bit redundant and everything should be under the main provider folder.
Yeah, the plan was ProviderV2 would be in the same packacge `qiskit.providers.v2` as `ProviderV1`. Sure, I can move it up a level, I was just worried about about having `basebackend.py` and `backend.py` in the same directory..

File: qiskit/providers/v2/provider.py
Comment (mtreinish) [reply qiskit/providers/v2/provider.py:34]: Done in: https://github.com/Qiskit/qiskit-terra/pull/5086/commits/0faaee159abf409558a40757ed44136cc65bc0bb

File: qiskit/providers/v2/job.py
Comment (mtreinish) [reply qiskit/providers/v2/job.py:46]: Done in: https://github.com/Qiskit/qiskit-terra/pull/5086/commits/0faaee159abf409558a40757ed44136cc65bc0bb

File: qiskit/providers/v2/backend.py
Comment (mtreinish) [reply qiskit/providers/v2/backend.py]: I updated it here: https://github.com/Qiskit/qiskit-terra/pull/5086/commits/523fba97ded10606489869d4f8392e70b7febf16 hopefully it makes more sense now

File: qiskit/providers/v2/backend.py
Comment (mtreinish) [reply qiskit/providers/v2/backend.py]: I reworded this in: https://github.com/Qiskit/qiskit-terra/pull/5086/commits/523fba97ded10606489869d4f8392e70b7febf16

File: qiskit/providers/v2/options.py
Comment (mtreinish) [reply qiskit/providers/v2/options.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/5086/commits/6533f3402a141aaecf3256e82493dee951b00e8d

**Comment by jyu00** [comment]
> > i'm a little confused by v1 vs. v2 - there's a v2 folder, is the intention that there will be v3, v4, etc. folders in the future? or by v2 you mean "everything that will be versioned from now on"?
> 
> The later, basically the v2 subdirectory was supposed to indicate the new version of the entire interface. In that new interface all the objects are individually versioned (starting at 1). I've been thinking of it as more of a `Backend` v2.1, etc. 

I understand the reasons behind this, but having `ProviderV1` being a "v2 provider" is a bit confusing. Maybe we can use a different word for `v1` and `v2` providers? `generation 2`? `model 2`?

File: qiskit/providers/v2/options.py
Comment (jyu00) [reply qiskit/providers/v2/options.py:29]: I think you still need a `@options.setter` in `Backend` to do `backend.options.shots`, since right now `_options` is the real attribute. Maybe just convert `set_options()` to a setter?

Comment (mtreinish) [reply qiskit/providers/v2/options.py:29]: I think it should be good in it's current form because there is a `@parameter` decorator on `options`. So `backend.options` will return the `Options` object and then `.shots =` will call setattr on that object. If we added a setter on options to `Backend` it would be called when `Backend.options =`. I do think I should try to figure out if there is a way to add a `__setattr__` method to this class to call `update_options` since right now you could call `Backend.options.secret_shots = 10` and that would get set even though its probably not a valid option. But, maybe that's not a problem

File: qiskit/providers/__init__.py
Comment (ajavadia) [inline qiskit/providers/__init__.py:25]:
```suggestion
objects. This module contains the abstract classes which are used to define the
```

File: qiskit/providers/job.py
Comment (ajavadia) [inline qiskit/providers/job.py:40]:
```suggestion
    This first version of the Backend abstract class is written to be mostly
    backwards compatible with the legacy providers interface. This was done to ease
    the transition for users and provider maintainers to the new versioned providers. Expect,
```

File: qiskit/providers/backend.py
Comment (ajavadia) [inline qiskit/providers/backend.py:55]:
```suggestion
    This first version of the Backend abstract class is written to be mostly
    backwards compatible with the legacy providers interface. This includes reusing
    the model objects :class:`~qiskit.providers.models.BackendProperties` and
    :class:`~qiskit.providers.models.BackendConfiguration`. This was done to
    ease the transition for users and provider maintainers to the new versioned providers.
```

File: qiskit/providers/backend.py
Comment (ajavadia) [inline qiskit/providers/backend.py:204]:
```suggestion
                For legacy providers migrating to the new versioned providers,
```

**Review by ajavadia — COMMENTED** [review]
Looks good, some comments inline to better reflect the terminology we agreed on. The release notes also needs updating to reflect the new directory layout.

**Review by ajavadia — APPROVED** [review]

**Pull request merged**
